pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+					Self::do_buy(offer.maker, class_id, instance_id, true)?;
+					Self::deposit_event(Event::OfferAccepted(sender, class_id, instance_id));
+					Ok(())
+				} else {
+					Err(Error::<T>::OfferExpired.into())
+				}
+			})
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The price for a token was updated \[owner, class_id, instance_id, price\]
+		TokenPriceUpdated(T::AccountId, T::NftClassId, T::NftInstanceId, Option<BalanceOf<T>>),
+		/// Token was sold to a new owner \[owner, buyer, class_id, instance_id, price, author, royalty, royalty_amount\]
+		TokenSold(
+			T::AccountId,
+			T::AccountId,
+			T::NftClassId,
+			T::NftInstanceId,
+			BalanceOf<T>,
+			Option<(T::AccountId, u8)>,
+			BalanceOf<T>,
+		),
+		/// Token listed on Marketplace \[owner, class_id, instance_id, author royalty\]
+		TokenListed(T::AccountId, T::NftClassId, T::NftInstanceId, T::AccountId, u8),
+		/// Token listed on Marketplace \[class_id, instance_id\]
+		TokenUnlisted(T::NftClassId, T::NftInstanceId),
+		/// Offer was placed on a token \[offerer, class_id, instance_id, price\]
+		OfferPlaced(T::AccountId, T::NftClassId, T::NftInstanceId, BalanceOf<T>),
+		/// Offer was withdrawn \[sender, class_id, instance_id\]
+		OfferWithdrawn(T::AccountId, T::NftClassId, T::NftInstanceId),
+		/// Offer was accepted \[sender, class_id, instance_id\]
+		OfferAccepted(T::AccountId, T::NftClassId, T::NftInstanceId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Account is not the owner of the token
+		NotTheTokenOwner,
+		/// Cannot buy a token from yourself
+		BuyFromSelf,
+		/// Token is currently not for sale
+		NotForSale,
+		/// This class type cannot be listed on Marketplace
+		UnsupportedClassType,
+		/// Token is not listed on Marketplace
+		NotListed,
+		/// Class or instance does not exist
+		ClassOrInstanceUnknown,
+		/// Token already listed on marketplace
+		AlreadyListed,
+		/// Offer is lower than the pallet limit
+		OfferTooLow,
+		/// No offer for this token found from the user
+		UnknownOffer,
+		/// Offer is no longer valid
+		OfferExpired,
+		/// User already offered for this token
+		AlreadyOffered,
+		/// User has to be offer maker or token owner to withdraw an offer
+		WithdrawNotAuthorized,
+		/// User has to be the token owner to accept an offer
+		AcceptNotAuthorized,
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	// Call extrinsic helper function used by `buy` and `accept_offer` functions
+	fn do_buy(
+		buyer: T::AccountId,
+		class_id: T::NftClassId,
+		instance_id: T::NftInstanceId,
+		is_offer: bool,
+	) -> DispatchResult {
+		let owner = pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into())
+			.ok_or(Error::<T>::ClassOrInstanceUnknown)?;
+		ensure!(buyer != owner, Error::<T>::BuyFromSelf);
+
+		let owner_origin = T::Origin::from(RawOrigin::Signed(owner.clone()));
+
+		pallet_uniques::Pallet::<T>::thaw(owner_origin, class_id.into(), instance_id.into())?;
```

Comment:
Move this inside the try_mutate too. 

so in case try_mutates returns Err - this will be reverted too.